changefeedccl: optimize enriched source provider

asg0451 opened the pull request:
Optimize enriched source provider by avoiding unnecessary
allocations and json construction.

Fixes: #141798

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142684)
<!-- Reviewable:end -->

asg0451 committed:
changefeedccl: optimize enriched source provider

Optimize enriched source provider by avoiding unnecessary
allocations and json construction.

Fixes: #141798

Release note: None

asg0451 commented:
the first commit in this PR is https://github.com/cockroachdb/cockroach/pull/142668 and can be ignored for the purpose of review

asg0451 commented on pkg/util/json/json.go:
```diff
@@ -458,6 +460,69 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
+func NewPartialObject(base map[string]JSON, newKeys []string) (*PartialObject, error) {
+	// Check that newKeys are unique.
+	newKeySet := make(map[string]struct{}, len(newKeys))
+	for _, k := range newKeys {
```
🤦

andyyang890 commented on pkg/util/json/json.go:
```diff
@@ -460,6 +462,70 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
+	// Check that newKeys are unique.
+	newKeySet := make(map[string]struct{}, len(newKeys))
+	for _, k := range newKeys {
+		if _, ok := newKeySet[k]; ok {
```
Do we also want to assert that the new keys aren't already in `base`?

asg0451 replied:
yes i think we do, good call

andyyang890 commented on pkg/util/json/json.go:
```diff
@@ -460,6 +462,70 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
+		newKeySet[k] = struct{}{}
+	}
+
+	sort.Strings(newKeys)
```
This sort isn't necessary given we're sorting again below after appending `newKeys`

andyyang890 commented on pkg/util/json/json.go:
```diff
@@ -460,6 +462,70 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
+// NewObject creates a new JSON object with the specified new key-values. The
+// order of newData MUST match the order of newKeys supplied to this
+// PartialObject's constructor.
+func (po *PartialObject) NewObject(newData []JSON) (JSON, error) {
```
I wonder if it would be less error prone to have this function accept a `map[string]JSON`

asg0451 replied:
yeah i was going back and forth on this myself. i'll change it, thx for the input

andyyang890 commented on pkg/util/json/json.go:
```diff
@@ -460,6 +462,70 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
 	return jsonObject(append([]jsonKeyValuePair(nil), b.pairs...)), nil
 }
 
+// PartialObject is a JSON object builder that builds an object using a set of
```
Should this be named `PartialObjectBuilder`?

andyyang890 replied:
`FixedKeysObjectBuilderWithPartiallyPrepopulatedValues` 😛

andyyang890 commented on pkg/ccl/changefeedccl/helpers_test.go:
```diff
@@ -1582,8 +1582,12 @@ func getTestingEnrichedSourceData() enrichedSourceData {
 // for use in tests.
-func getTestingEnrichedSourceProvider(opts changefeedbase.EncodingOptions) *enrichedSourceProvider {
-	return newEnrichedSourceProvider(opts, getTestingEnrichedSourceData())
+func getTestingEnrichedSourceProvider(
```
nit: would prefer the name of this function began with `new` instead of `get` given it's creating a new one instead of fetching one that already exists

asg0451 replied:
i dont disagree but dont want to change it here, since i didnt introduce this method and dont want to inflate the pr even more

andyyang890 commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -943,21 +943,26 @@ func (e EncodingOptions) Validate() error {
 			if v.b {
-				return errors.Errorf(`%s is only usable with %s=%s`,
-					v.k, OptEnvelope, OptEnvelopeWrapped)
+				return errors.Errorf(`%s is only usable with %s=%s or %s`,
```
nit:
```suggestion
				return errors.Errorf(`%s is only usable with "%[2]s=%s or %[2]s=%s"`,
```

asg0451 replied:
what does %[2]s do?

andyyang890 replied:
yeah you can, i think you can think of each raw `%s` as if it was the next `%[n]s` so
```
`%s is only usable with "%[2]s=%s or %[2]s=%s"`
```
is equivalent to
```
`%[1]s is only usable with "%[2]s=%[3]s or %[2]s=%[4]s"`,
```

andyyang890 commented on pkg/ccl/changefeedccl/enriched_source_provider.go:
```diff
@@ -87,76 +91,81 @@ func newEnrichedSourceData(
 		sourceData: sourceData,
 		opts: enrichedSourceProviderOpts{
 			mvccTimestamp: opts.MVCCTimestamps,
 			updated:       opts.UpdatedTimestamps,
```
`updated` appears to be unused. Could you add a TODO with an issue link to make sure we use it?

asg0451 replied:
this is covered by these TODOs: `// TODO(#139661): Add other non fixed fields, in the right order.`

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -10748,3 +10770,9 @@ func TestChangefeedProtectedTimestampUpdate(t *testing.T) {
 	cdcTest(t, testFn, feedTestForceSink("kafka"), withTxnRetries)
 }
+
+func assertReasonableMVCCTs(t *testing.T, ts string) {
```
super nit: `Ts` should be capitalized like `TS` but given that would make it look like one long acronym, maybe `MVCCTimestamp`?

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -4236,96 +4237,117 @@ func TestChangefeedEnrichedSourceWithData(t *testing.T) {
-							nodeID = actualSource["node_id"]
-						}
-						require.NotNil(t, nodeID)
+		for _, withMVCCTs := range []bool{true, false} {
```
You can use `testutils.RunTrueAndFalse` instead, that would also get you the `mvcc_ts=%v` subtest naming for free

asg0451 replied:
nice

andyyang890 commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -4236,96 +4237,117 @@ func TestChangefeedEnrichedSourceWithData(t *testing.T) {
+
+						sourceAssertion = func(actualSource map[string]any) {
+							var nodeID any
+							if testCase.format == "avro" {
```
nit: use a switch statement instead and explicitly label the `"json"` case

asg0451 replied:
is that really meaningfully better?

andyyang890 replied:
I think it makes it slightly more readable heh, you could also add a default case that fails the test due to an unexpected format but that might be overkill

asg0451 replied:
ehhhh i think this is fine how it is ;) ;)

asg0451 commented on pkg/util/json/json_test.go:
```diff
@@ -744,6 +744,44 @@ func TestBuildFixedKeysJSONObjectErrors(t *testing.T) {
+	}
+	newKeys := []string{"c", "c"}
+	_, err := NewPartialObject(base, newKeys)
+	require.Error(t, err)
```
why? i dont like to assert on error messages unless they're user facing, or depended on by code which predates errors.Is

andyyang890 replied:
I was thinking mostly to make sure we're hitting the right error. I've seen instances in the past where we've had checks like this and then after a code change, a different error than what we expected was being returned and we would've caught it sooner had the test been more specific. I realize it's unlikely to be the case here but I think it'd still be a nice check to have.

andyyang890 commented on pkg/util/json/json.go:
```diff
@@ -460,6 +462,82 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
+
+	// Check that newKeys are disjoint from base keys.
+	for k := range base {
+		if _, ok := newKeySet[k]; ok {
```
nit: you can move this check into the same loop as above, i.e. check if the key is in `base`

andyyang890 commented on pkg/util/json/json.go:
```diff
@@ -460,6 +462,82 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
+			return nil, errors.AssertionFailedf("unknown new key %s", k)
+		}
+		if _, ok := po.keyOrd[k]; !ok {
+			return nil, errors.AssertionFailedf("unknown key %s", k)
```
nit: might be more accurate to say "missing key ord"

asg0451 replied:
the meaning of a missing key ord is that you provided a key that wasnt in the constructor, so i think the message is ok

andyyang890 replied:
But shouldn't that materialize already by failing the previous check? If it's in `allowedNewKeys` but isn't in `keyOrd`, wouldn't that imply we made a mistake and that it's a valid new key but we forgot to assign it a `keyOrd`?

andyyang890 commented on pkg/util/json/json.go:
```diff
@@ -460,6 +462,70 @@ func (b *FixedKeysObjectBuilder) Build() (JSON, error) {
+// PartialObject is a JSON object builder that builds an object using a set of
+// fixed key-values (supplied at construction) and a set of keys which are
+// settable when building the final JSON object.
+type PartialObject struct {
```
True ok, when I was skimming the comments I didn't notice that `Build()` resets all the keys, I thought it preserved all the values and let you build again after only changing a couple.

asg0451 commented:
bors r=andyyang890
